### PR TITLE
UPDATE /profile/address endpoint

### LINKED ISSUE #59: UPDATE or add a new address model to user
- Allow user to update current addreses
- Allow user to add more than one address
- Allow user to add a new address to the address list

### PR DESCRIPTION
**Pull request by awiednoor**
- User can add a new address to address array
- User can delete an existing address
- User can update an existing address
- Address is validated before updation

### CODE REVIEW

**Review by khatibAmjad — APPROVED** [review]

File: src/controllers/user.js [hunk 1]
```diff
@@ -138,6 +128,37 @@ const updatePassword = async (req, res) => {
   }
 };
 
+const updateAddress = async (req, res) => {
+  const User = await UserModel.findOne({ _id: req.user.userId });
```
Comment (halitbatur) [inline src/controllers/user.js]: okay the variable name here shoulnd't be capitalized it should be camelCased

**Review by halitbatur — APPROVED** [review]

File: src/controllers/user.js [hunk 2]
```diff
@@ -138,6 +128,37 @@ const updatePassword = async (req, res) => {
...
+    const newAddress = req.body.address;
+    let message;
+    if (addressID && operation) {
+      const toUpdateAddress = User.address.filter(
+        (address) => String(address._id) === addressID
+      );
+      const index = User.address.indexOf(toUpdateAddress[0]);
+      if (index === -1)
+        return res.status(404).json({ message: 'Address not found' });
+      if (operation === 'update') {
```
Comment (khaldarov) [inline src/controllers/user.js]: I tried to update a current address, and here's the outcome:
- This was the address we already had:
Screenshot from 2022-08-18 17-45-48
- I tried to change it, and this is what I sent:
Screenshot from 2022-08-18 17-45-51
- I received a "address added," but it was added as a `null`, let alone the fact that it should've been an update.
- And this is what one receives when sends in the example:
Screenshot from 2022-08-18 17-47-41

**Review by khaldarov — CHANGES_REQUESTED** [review]

**Review by khaldarov — CHANGES_REQUESTED** [review]
- The docs are missing the `/api/user` part.
- A `title` field _could_ be added to the `addressSchema` to make the errors or messages easier to show. `title` might be "home" or "work," maybe. I should've done that already, sorry for that. Following that,
- When an address is deleted, it might show a `${title} address has be deleted`

### According to the documentation
- it doesn't add an address,
- it doesn't delete an address:
Screenshot from 2022-08-19 12-40-39
- it doesn't update an address:
Screenshot from 2022-08-19 12-42-10

### Its format
- should not be like this:
Screenshot from 2022-08-19 12-43-17
- but should be like this:
Screenshot from 2022-08-19 12-43-29

It's taking a lot of time, I assume. Also, I can work on this. We can leave it for a later time if you wish.

**Review by khaldarov — COMMENTED** [review]
Okay so after a test on Postman:
- Adding an address this way (which is the correct way), giving the correct message only adds a null:
Screenshot from 2022-08-19 13-05-16
- Adding an address in this format does work:
Screenshot from 2022-08-19 13-07-44
- Deleting that address doesn't work:
Screenshot from 2022-08-19 13-06-59

**Comment by awiednoor** [comment]
All requested changes were made, enhanced controller and fixed documentation.
Test one last time and if everything is working fine we can merge

File: src/controllers/user.js [hunk 3]
```diff
@@ -90,17 +90,7 @@ const getSingleUser = async (req, res) => {
...
-    'age',
-    'gender',
-    'nationality',
-    'refugee',
-    'profilePhoto',
-  ];
+  const allowedParams = constants.VALID_USER_KEYS;
   const emptyOrWhiteSpace = /^\s*$/;
   try {
     const User = await UserModel.findById(req.user.userId);
```
Comment (khaldarov) [inline src/controllers/user.js:94]: `User` should be camelCased in ALL files; I'm adding this upon Halit's previous request.

File: src/controllers/user.js [hunk 4]
```diff
@@ -138,6 +128,45 @@ const updatePassword = async (req, res) => {
...
 
+const updateAddress = async (req, res) => {
+  try {
+    const { addressID, operation } = req.params;
+    const User = await UserModel.findById(req.user.userId);
+    let message;
+    if (operation === 'add') {
+      if (req.body.address !== null && req.body.address !== undefined) {
+        User.address.push(req.body.address);
+        message = `${req.body.address.title} Address added`;
```
Comment (khaldarov) [inline src/controllers/user.js]: `Address` should also be in lowercase. Maybe a "... address has been added"

File: src/controllers/user.js [hunk 5]
```diff
@@ -138,6 +128,45 @@ const updatePassword = async (req, res) => {
...
+    } else if (operation === 'delete' || operation === 'update') {
+      if (addressID) {
+        const addressIndex = User.address
+          .map((address) => address.id)
+          .indexOf(addressID);
+        if (addressIndex === -1)
+          return res.status(404).json({ message: 'Address not found' });
+        if (operation === 'update') {
+          User.address[addressIndex] = req.body.address;
+          message = ` ${User.address[addressIndex].title} address has been updated`;
```
Comment (khaldarov) [inline src/controllers/user.js]: The space at the very first column should be deleted.

File: src/docs/User/updateProfile.yaml [hunk 6]
```diff
@@ -122,3 +122,90 @@ paths:
...
+    patch:
+      tags:
+        - Update User
+      summary: User can update an address in the current address list, can delete and add new addresses to the list
+      description: API has 2 parameters , first is the address ID in the address list and the operation which can have one of three values [add,update,delete], in address object [title, country, city, address0] are required
+      parameters:
+        - name: operation
+          in: path
+          required: true
+          schema:
```
Comment (khaldarov) [inline src/docs/User/updateProfile.yaml:135]: We might add an enum to make it easier to read.

File: src/docs/User/updateProfile.yaml [hunk 7]
```diff
@@ -122,3 +122,90 @@ paths:
...
+                  }
+                  }'
+              'ADDRESS IS NOT IN A VALID FORMAT':
+                value: '{
+                  "address" : "Turkey, Istanbul"
+                  }'
+              'ONE REQUIRED FIELD IS MISSING':
+                value: '{
+                  "address" : {
+                  "title": "Home",
```
Comment (khaldarov) [inline src/docs/User/updateProfile.yaml:187]: They do work, but we might add a `"Country": "",` empty string here to make sure they work.

File: src/middleware/updateValidation.js [hunk 8]
```diff
@@ -65,7 +65,40 @@ const validatePassword = async (req, res, next) => {
...
+    return res.status(400).json({ error: 'Invalid operation' });
+  }
+  if (address) {
+    if (typeof address !== 'object') {
+      errorsArray.push('Address is not an object');
+    }
+    // eslint-disable-next-line no-restricted-syntax
+    for (const field of requiredFeilds) {
+      if (!address[field]) {
+        errorsArray.push(`${field} is required in Address`);
```
Comment (khaldarov) [inline src/middleware/updateValidation.js]: `Address` should be lowercased.

File: src/controllers/user.js [hunk 9]
```diff
@@ -138,6 +128,45 @@ const updatePassword = async (req, res) => {
   }
 };
 
+const updateAddress = async (req, res) => {
+  try {
+    const { addressID, operation } = req.params;
+    const User = await UserModel.findById(req.user.userId);
+    let message;
+    if (operation === 'add') {
+      if (req.body.address !== null && req.body.address !== undefined) {
```
Comment (khaldarov) [inline src/controllers/user.js:134]: We can improve this later to extract the address directly from the `req.body`, and not from the `req.body.address`.

**Review by khaldarov — APPROVED** [review]
Very well done, Noor. It looks great.
- Screenshot from 2022-08-20 12-36-01
- Screenshot from 2022-08-20 12-36-40

**Pull request merged**